### ISSUE TITLE
Reliance on gRPC ordering in consensus

### ISSUE THREAD

**Issue by jyellick** [hyperledger-archives/fabric#2053]
## Description

The gRPC protocol guarantees us that, for a single stream, messages will be delivered in order.

The PBFT protocol assumes that message transmission is 'reliable', but not that message transmission is in order.  The current PBFT implementation assumes this weaker guarantee, but there could be significant simplifications and optimizations if we were to exploit the gRPC delivery order guarantee.

My question to the architects and others interested in the high level project design, is whether it is safe for consensus to rely on the ordering guarantee of gRPC? Or, is it possible that we will be switching off gRPC to some transport which does not preserve the ordering guarantee, and we should therefore continue to attempt to keep the PBFT implementation agnostic to message ordering?

I've discussed this briefly with @kchristidis and our rough conclusion was, it would be nice if we could simplify our lives by relying on gRPC, but we need to decide concretely as it will affect many of the consensus architectural decisions going forward.

@binhn @ghaskins @srderson @jeffgarratt @vukolic @cca88 @tuand27613 @corecode

**Comment by vukolic** [comment]
Protocol-wise one cannot really  leverage FIFO channels to simplify, e.g., PBFT 3-phase common-case communication pattern. 

Did you have some other usage for FIFO gRPCs in mind?

**Comment by mastersingh24** [comment]
Even for something like atomic broadcast you could not solely rely on gRPC since I don't believe that the protocol provides any type of sequence number (or at least not that I have seen).  Last time I looked, maintaining order is handled by locking access to the payload (done of the clients only have a single thread for handling messages for any given stream to prevent concurrent access).  Now if you added a sequence number and kept track of the last one processed, you could potentially avoid dealing with reordering on the "deliver" call.

**Comment by kchristidis** [comment]
Jason made the gRPC observation and will expand on this more, but roughly: Messages that are always delivered in order means we do not have to deal with a _window_ of possibly valid messages (i.e. "anything between seqNo 11 and 18 is good"), and instead make the criterion for the next expected PBFT message from a replica pretty simple. An example that @jyellick used was: "if the next pre-prepare that you receive from the primary doesn't correspond to seqNo 11 (i.e. it creates a hole), you should view-change". So you can argue that this would allow you to detect irregularities much faster. If you think of it as a state machine, a FIFO link means that the set of transitions that you need to consider is smaller/tighter.

**Comment by mastersingh24** [comment]
That's what I was saying as well.  I assume the seq numbers are in the messages

**Comment by kchristidis** [comment]
They are. I wasn't sure what you meant by "avoid dealing with reordering on the deliver call" which is why I hesitated to agree with the original statement.

**Comment by cca88** [comment]
Indeed, this is a useful idea. It does simplify the protocol, even if it only means that there is no FIFO ordering layer needed "on top", as I have seen discussed elsewhere. (Note: atomic broadcast with FIFO order means that msgs are by-sender FIFO-ordered.)
Back then we had implemented SINTRA over TCP because it was the right thing to do, that is, with FIFO ordered reliable links. PBFT and everyone in that camp thought TCP was too slow and went for UDP, which obviously is wrong for a WAN.
If FIFO comes in, I would only do it on top of a reliable links guarantee, that the sender ensures a receiver doesnt miss any message. This means you replace the window mechanism with a presumably more complex by-receiver buffering scheme and so on.
Another question: does gRPC have some notion of scope inside which it does FIFO but outside not? Like, what happens if the underlying TCP connection dies? could this break FIFO? Would this require keeping any persistent state on the sender/receiver?
I wouldnt change this for now, enough work has gone into the window mechanism.

**Comment by jyellick** [comment]
@vukolic @cca88 I would not advocate immediately abandoning the windowed concept, as you're correct, we have put a fair amount of effort into making this work.

However, as @kchristidis pointed out, there are still a number of optimizations we could make which would make the implementation both more robust and simpler for future features.  The prePrepare no holes ordering is an obvious one, but there are a couple others which jump to mind.
1. Take the discussion we had on slack for instance.  In order to have multiple requests in flight, without ordering, we must track each request individually, and make careful decisions about whether the request we are receiving has already been processed by the network.  Without message ordering, a request with a later timestamp (or seqno) may arrive after a later request has executed, and without careful accounting, it's not obvious whether that request is stale or new.
2. Similarly, with a replica detecting it is out of sync with the network, if it receives 2f+1 prepare messages for seqNo=12, but it has f or fewer prepare messages for seqNo=11, then it can confidently assert that it will never have 2f+1 prepares for seqNo=11, and should attempt to recover via state transfer.

My understanding is that if the gRPC link is broken, then a new handler will be created for the new link, and we could write a hook to alert us to this condition.  @jeffgarratt might know more about link failure.

Finally, to your point @cca88 

> ... went for UDP, which obviously is wrong for a WAN.

I think that is really the crux of this issue.  I'm not ready to assert that UDP is an obviously wrong choice here.  As you point out, TCP is reliable and ordered, except under weird conditions where it's not (like when links break and must be re-established).  I think there are those who would say that anything implemented in TCP can be implemented over UDP more efficiently.  I see both sides of this debate, but am really looking for architectural guidance as to what the hyperledger fabric's long term direction is.  If we intend to pursue less reliable links in the future, then by all means, let's not introduce any dependencies on the gRPC/TCP ordering, but if this is not on the table, it would seem foolish not to take advantage of the optimizations and simplifications allowed by the ordering.

**Comment by kchristidis** [comment]
> My understanding is that if the gRPC link is broken, then a new handler will be created for the new link, and we could write a hook to alert us to this condition. @jeffgarratt might know more about link failure.

(I've dealt with that piece of code when building the revised discovery service.) This statement is true, we can easily tell when a link is established or broken.

**Comment by vukolic** [comment]
The decision whether to leverage gRPC FIFO to simplify consensus needs to be preceded by the decision whether we want gRPC for intra consensus service communication in the first place, once we separate consensus service from the peers.

**Comment by binhn** [comment]
If we need to change, July is the time. We are using gRPC stream, which guarantees FIFO.

**Comment by thiruworkspace** [comment]
By relying on gRPC for FIFO/ordering, it may short term simply the model, but it would limit the channel protocol options for future.

it will be good if someone can highlight the documentation on recovery logic when the application breaks or network breaks?
